src/hotspot/share/opto/superword.cpp

```diff
@@ -1214,110 +1186,136 @@ bool SuperWord::follow_def_uses(Node_List* p) {
     }
-    return true;
+    return PairOrderStatus::Ordered;
   }
 
-  uint ct = u1->req();
-  if (ct != u2->req()) return false;
+  uint ct = use1->req();
+  if (ct != use2->req()) { return PairOrderStatus::Unordered; };
   uint i1 = 0;
   uint i2 = 0;
   do {
-    for (i1++; i1 < ct; i1++) if (u1->in(i1) == d1) break;
-    for (i2++; i2 < ct; i2++) if (u2->in(i2) == d2) break;
+    for (i1++; i1 < ct; i1++) { if (use1->in(i1) == def1) { break; } }
+    for (i2++; i2 < ct; i2++) { if (use2->in(i2) == def2) { break; } }
     if (i1 != i2) {
-      if ((i1 == (3-i2)) && (u2->is_Add() || u2->is_Mul())) {
-        // Further analysis relies on operands position matching.
-        u2->swap_edges(i1, i2);
-      } else if (VectorNode::is_muladds2i(u2) && u1 != u2) {
+      if ((i1 == (3-i2)) && (use2->is_Add() || use2->is_Mul())) {
+        // 3. Add/Mul: swap edges, and hope the other position matches too.
+        use2->swap_edges(i1, i2);
+      } else if (VectorNode::is_muladds2i(use2) && use1 != use2) {
+        // 4.a/b: MulAddS2I.
         if (i1 == 5 - i2) { // ((i1 == 3 && i2 == 2) || (i1 == 2 && i2 == 3) || (i1 == 1 && i2 == 4) || (i1 == 4 && i2 == 1))
-          u2->swap_edges(1, 2);
-          u2->swap_edges(3, 4);
+          use2->swap_edges(1, 2);
+          use2->swap_edges(3, 4);
         }
         if (i1 == 3 - i2 || i1 == 7 - i2) { // ((i1 == 1 && i2 == 2) || (i1 == 2 && i2 == 1) || (i1 == 3 && i2 == 4) || (i1 == 4 && i2 == 3))
-          u2->swap_edges(2, 3);
-          u2->swap_edges(1, 4);
+          use2->swap_edges(2, 3);
+          use2->swap_edges(1, 4);
         }
-        return false; // Just swap the edges, the muladds2i nodes get packed in follow_use_defs
+        return PairOrderStatus::Unknown;
       } else {
-        return false;
+        // The inputs are not ordered, and we can not do anything about it.
+        return PairOrderStatus::Unordered;
       }
-    } else if (i1 == i2 && VectorNode::is_muladds2i(u2) && u1 != u2) {
-      u2->swap_edges(1, 3);
-      u2->swap_edges(2, 4);
-      return false; // Just swap the edges, the muladds2i nodes get packed in follow_use_defs
+    } else if (i1 == i2 && VectorNode::is_muladds2i(use2) && use1 != use2) {
+      // 4.c: MulAddS2I.
+      use2->swap_edges(1, 3);
+      use2->swap_edges(2, 4);
+      return PairOrderStatus::Unknown;
     }
   } while (i1 < ct);
-  return true;
+
+  // 2. All inputs match.
+  return PairOrderStatus::Ordered;
 }
 
-//------------------------------est_savings---------------------------
 // Estimate the savings from executing s1 and s2 as a pack
-int SuperWord::est_savings(Node* s1, Node* s2) {
+int SuperWord::estimate_cost_savings_when_packing_pair(const Node* s1, const Node* s2) const {
   int save_in = 2 - 1; // 2 operations per instruction in packed form
 
+  auto adjacent_profit = [&] (Node* s1, Node* s2) { return 2; };
+  auto pack_cost       = [&] (int ct) { return ct; };
+  auto unpack_cost     = [&] (int ct) { return ct; };
```

Comment:
Note: I moved the methods from the `SuperWord` class to lambdas in this method, they were not used anywhere else.